crates/interpreter/src/frame.rs

```diff
@@ -0,0 +1,176 @@
+use std::mem;
+
+use cranelift_entity::SecondaryMap;
+
+use sonatina_ir::{
+    module::ModuleCtx,
+    types::{CompoundType, CompoundTypeData},
+    DataFlowGraph, Function, Insn, Type, Value, ValueData, I256,
+};
+
+use crate::{Literal, ProgramCounter};
+
+pub struct Frame {
+    pub ret_addr: ProgramCounter,
+    local_values: SecondaryMap<Value, Literal>, // 256-bit register
+    alloca_region: Vec<u8>,                     // big endian
+}
+
+impl Frame {
+    pub fn new(func: &Function, ret_addr: ProgramCounter, args: Vec<Literal>) -> Self {
+        let mut local_values = SecondaryMap::new();
+        for (v, literal_value) in func.arg_values.iter().zip(args.into_iter()) {
+            local_values[*v] = literal_value
+        }
+        let alloca_region = Vec::new();
+
+        Self {
+            ret_addr,
+            local_values,
+            alloca_region,
+        }
+    }
+
+    pub fn load(&mut self, /*ctx: Context,*/ v: Value, dfg: &DataFlowGraph) -> Literal {
+        if !self.is_assigned(v) {
+            let v = match dfg.value_data(v) {
+                ValueData::Insn { insn, .. } => {
+                    let result_v = dfg.insn_result(*insn).unwrap();
+                    if self.is_assigned(result_v) {
+                        return self.local_values[result_v];
+                    }
+                    result_v
+                }
+                _ => v,
+            };
+            let i256 = dfg.value_imm(v).unwrap().as_i256();
+            self.local_values[v] = Literal(i256);
+        }
+        self.local_values[v]
+    }
+
+    pub fn map(&mut self, literal: Literal, insn: Insn, dfg: &DataFlowGraph) {
+        let v = dfg.insn_result(insn).unwrap();
+        debug_assert!(!self.is_assigned(v));
+        self.local_values[v] = literal
+    }
+
+    pub fn alloca(&mut self, ctx: &ModuleCtx, ty: Type, insn: Insn, dfg: &DataFlowGraph) {
```

Comment:
```suggestion
    pub fn alloca(&mut self, ctx: &ModuleCtx, ty: Type) -> LIteral
```
How about just returning a literal value that points to the allocated region?
Related to the aforementioned comment, it'd be preferable for the frame to be independent of the concept of insn`.